packages/discovery-provider/plugins/pedalboard/apps/solana-relay/src/middleware/rateLimiter.ts

```diff
@@ -0,0 +1,67 @@
+import { Request, Response, NextFunction } from "express"
+import { RedisClientType } from "redis"
+import { getIP } from "../utils/ipData"
+import { getRedisConnection } from "../redis"
+import { config } from "../config"
+
+export class RateLimiter {
+    private hourlyLimit: number
+    private dailyLimit: number
+    private weeklyLimit: number
+    private prefix: string
+
+    constructor(prefix: string, hourlyLimit: number, dailyLimit: number, weeklyLimit: number) {
+        this.prefix = prefix
+        this.hourlyLimit = hourlyLimit
+        this.dailyLimit = dailyLimit
+        this.weeklyLimit = weeklyLimit
+    }
+
+    private getRedisKey(ip: string, period: string): string {
+        return `${this.prefix}:${ip}:${period}`
+    }
+
+    private async incrementCount(ip: string, period: string, limit: number, expiry: number): Promise<boolean> {
+        const key = this.getRedisKey(ip, period)
+        const redis = await getRedisConnection()
+        const currentCount = await redis.incr(key)
+        if (currentCount === 1) {
+            await redis.expire(key, expiry)
+        }
+        return currentCount >= limit
+    }
+
+    public async checkLimit(ip: string): Promise<{
+        hourLimitReached: boolean,
+        dayLimitReached: boolean,
+        weekLimitReached: boolean,
+        allowed: boolean
+    }> {
+        const hourLimitReached = await this.incrementCount(ip, 'hour', this.hourlyLimit, 3600)
+        const dayLimitReached = await this.incrementCount(ip, 'day', this.dailyLimit, 86400)
```

Comment:
promise all

Reply:
Legit question: Do we want to full on fail this whole thing if we fail to increment any of these counts?
Both this form and `Promise.all` will error out if any of them fails. I assume we want that to be true, just confirming the intended logic.